replica/table.cc

```diff
@@ -542,26 +543,92 @@ void table::enable_off_strategy_trigger() {
     do_update_off_strategy_trigger();
 }
 
-compaction_group_vector table::make_compaction_groups() {
-    compaction_group_vector ret;
-    auto&& ranges = dht::split_token_range_msb(_x_log2_compaction_groups);
-    ret.reserve(ranges.size());
-    tlogger.debug("Created {} compaction groups for {}.{}", ranges.size(), _schema->ks_name(), _schema->cf_name());
-    size_t i = 0;
-    for (auto&& range : ranges) {
-        ret.emplace_back(std::make_unique<compaction_group>(*this, i++, std::move(range)));
+class single_compaction_group_manager final : public compaction_group_manager {
+    replica::table& _t;
+public:
+    single_compaction_group_manager(replica::table& t) : _t(t) {}
+
+    compaction_group_vector make_compaction_groups() const override {
+        compaction_group_vector r;
+        r.push_back(std::make_unique<compaction_group>(_t, size_t(0), dht::token_range::make_open_ended_both_sides()));
+        return r;
     }
-    return ret;
+    size_t compaction_group_of(dht::token) const override {
+        return 0;
+    }
+    size_t log2_compaction_groups() const override {
+        return 0;
+    }
+};
+
+class tablet_compaction_group_manager final : public compaction_group_manager {
+    replica::table& _t;
+private:
+    const locator::effective_replication_map_ptr& erm() const {
+        return _t.get_effective_replication_map();
+    }
+
+    const schema_ptr& schema() const {
+        return _t.schema();
+    }
+
+    const locator::tablet_map& tablet_map() const {
+        if (!_t.uses_tablets()) {
+            throw std::runtime_error(format("Tablet map cannot be retrieved for table {}.{} which tablet isn't enabled.", schema()->ks_name(), schema()->cf_name()));
```

Comment:
"which tablet isn't enabled" doesn't sound grammatical, did you mean "for which tablets are not enabled" ?

Reply:
This check doesn't seem necessary, because we would fail anyway inside get_tablet_map() in this case.

Reply:
It's a left over from previous version where I tried to avoid this abstraction for locating compaction group from token, but ended up switching back as ifs will be sprinkled the more code we add, that will only apply to tablet-enabled tables. Will remove in next version. thanks.